service/mapper/transitions_test.go

```diff
@@ -0,0 +1,1008 @@
+			HasFunc: func(_ flow.StateCommitment) bool {
+				return true
+			},
+			SizeFunc: func() uint {
+				return 42
+			},
+		}
+
+		tr, st := baselineFSM(t, StatusUpdating)
+		st.forest = forest
+
+		err := tr.UpdateTree(st)
+
+		assert.NoError(t, err)
+		assert.Equal(t, StatusMatched, st.status)
+	})
+
+	t.Run("handles invalid status", func(t *testing.T) {
+		t.Parallel()
+
+		tr, st := baselineFSM(t, StatusEmpty)
+
+		err := tr.UpdateTree(st)
+
+		assert.Error(t, err)
+	})
+
+	t.Run("handles feeder update failure", func(t *testing.T) {
+		t.Parallel()
+
+		forest := &mocks.Forest{
+			HasFunc: func(_ flow.StateCommitment) bool {
+				return false
+			},
+		}
+		feed := &mocks.Feeder{
+			UpdateFunc: func() (*ledger.TrieUpdate, error) {
+				return nil, mocks.DummyError
+			},
+		}
+
+		tr, st := baselineFSM(t, StatusUpdating)
+		st.forest = forest
+		tr.feed = feed
+
+		err := tr.UpdateTree(st)
+
+		assert.Error(t, err)
+	})
+
+	t.Run("handles forest parent tree not found", func(t *testing.T) {
+		t.Parallel()
+
+		forest := &mocks.Forest{
+			TreeFunc: func(_ flow.StateCommitment) (*trie.MTrie, bool) {
+				return nil, false
+			},
+			HasFunc: func(_ flow.StateCommitment) bool {
+				return false
+			},
+		}
+
+		tr, st := baselineFSM(t, StatusUpdating)
+		st.forest = forest
+
+		err := tr.UpdateTree(st)
+
+		assert.NoError(t, err)
+	})
+}
+
+func TestTransitions_CollectRegisters(t *testing.T) {
+	t.Run("nominal case", func(t *testing.T) {
+		t.Parallel()
+
+		testNextCommit := flow.StateCommitment{
+			0x2a, 0x2a, 0x2a, 0x2a, 0x2a, 0x2a, 0x2a, 0x2a,
+			0x2a, 0x2a, 0x2a, 0x2a, 0x2a, 0x2a, 0x2a, 0x2a,
+			0x2a, 0x2a, 0x2a, 0x2a, 0x2a, 0x2a, 0x2a, 0x2a,
+			0x2a, 0x2a, 0x2a, 0x2a, 0x2a, 0x2a, 0x2a, 0x2a,
+		}
+		testLastCommit := flow.StateCommitment{
+			0x2b, 0x2b, 0x2b, 0x2b, 0x2b, 0x2b, 0x2b, 0x2b,
+			0x2b, 0x2b, 0x2b, 0x2b, 0x2b, 0x2b, 0x2b, 0x2b,
+			0x2b, 0x2b, 0x2b, 0x2b, 0x2b, 0x2b, 0x2b, 0x2b,
+			0x2b, 0x2b, 0x2b, 0x2b, 0x2b, 0x2b, 0x2b, 0x2b,
+		}
+
+		testTrie, err := trie.NewMTrie(testRoot)
+		require.NoError(t, err)
+
+		forest := &mocks.Forest{
+			HasFunc: func(_ flow.StateCommitment) bool {
+				return true
+			},
+			TreeFunc: func(_ flow.StateCommitment) (*trie.MTrie, bool) {
+				return testTrie, true
+			},
+			PathsFunc: func(_ flow.StateCommitment) ([]ledger.Path, bool) {
+				return []ledger.Path{testPath1, testPath2, testPath3, testPath4, testPath5, testPath6}, true
+			},
+			ParentFunc: func(commit flow.StateCommitment) (flow.StateCommitment, bool) {
+				assert.Equal(t, testNextCommit, commit)
+
+				return testLastCommit, true
+			},
+		}
+		feed := &mocks.Feeder{
+			UpdateFunc: func() (*ledger.TrieUpdate, error) {
+				return testUpdate, nil
+			},
+		}
+
+		tr, st := baselineFSM(t, StatusMatched)
+		st.registers = make(map[ledger.Path]*ledger.Payload)
+
+		tr.feed = feed
+		st.forest = forest
+
+		err = tr.CollectRegisters(st)
+
+		assert.NoError(t, err)
+		assert.NotEmpty(t, st.registers)
```

Comment:
I would want to check whether the collected registers correspond to all of the desired steps, too?